src/util/hashFetch/index.js

```diff
@@ -0,0 +1,247 @@
+
+const fs = require('fs');
+const util = require('util');
+const path = require('path');
+
+import unzip from 'unzipper';
+
+import { getHashFetchPath } from '../host';
+import ExpoRetry from './expoRetry';
+import Contracts from '@parity/shared/lib/contracts';
+import { sha3 } from '@parity/api/lib/util/sha3';
+import { bytesToHex } from '@parity/api/lib/util/format';
+import { ensureDir as fsEnsureDir, emptyDir as fsEmptyDir } from 'fs-extra';
+import { http, https } from 'follow-redirects';
+
+const fsExists = util.promisify(fs.stat);
+const fsStat = util.promisify(fs.stat);
+const fsRename = util.promisify(fs.rename);
+const fsReadFile = util.promisify(fs.readFile);
+const fsUnlink = util.promisify(fs.unlink);
+const fsReaddir = util.promisify(fs.readdir);
+const fsRmdir = util.promisify(fs.rmdir);
+
+const MAX_DOWNLOADED_FILE_SIZE_BYTES = 10485760; // 20MB
+
+function registerFailedAttemptAndThrow (hash, url, e) {
+  ExpoRetry.get().registerFailedAttempt(hash, url);
+  throw e;
+}
+
+function checkHashMatch (hash, path) {
+  return fsReadFile(path).then(content => {
+    if (sha3(content) !== `0x${hash}`) { throw new Error(`Hashes don't match: expected 0x${hash}, got ${sha3(content)}`); }
+  });
+}
+
+// List directory contents and tell if each item is a file or a directory
+function ls (folderPath) {
+  return fsReaddir(folderPath)
+    .then(filenames =>
+      Promise.all(filenames.map(filename => {
+        const filePath = path.join(folderPath, filename);
+
+        return fsStat(filePath).then(stat => ({ isDirectory: stat.isDirectory(), filePath, filename }));
+      }))
+    );
+}
+
+function rawUnzipTo (zipPath, extractPath) {
+  return new Promise((resolve, reject) => {
+    const unzipParser = unzip.Extract({ path: extractPath });
+
+    fs.createReadStream(zipPath).pipe(unzipParser);
+
+    unzipParser.on('error', reject);
+
+    unzipParser.on('close', resolve);
+  });
+}
+
+function unzipThroughTo (tempPath, extractPath, finalPath) {
+  return rawUnzipTo(tempPath, extractPath)
+    .then(() => fsUnlink(tempPath))
+    .then(() => ls(extractPath))
+    .then(files => {
+      // Check if the zip file contained a root folder
+      if (files.length === 1 && files[0].isDirectory) {
+        // Rename the root folder (contaning the dapp) to finalPath
+        const rootFolderPath = files[0].filePath;
+
+        return fsRename(rootFolderPath, finalPath)
+            .then(() => fsRmdir(extractPath));
+      } else {
+        // No root folder: extractPath contains the dapp
+        return fsRename(extractPath, finalPath);
+      }
+    });
+}
+
+function download (url, destinationPath) {
+  const file = fs.createWriteStream(destinationPath); // Will replace any existing file
+
+  return new Promise((resolve, reject) => {
+    const httpx = url.startsWith('https:') ? https : http;
+
+    httpx.get(url, response => {
+      var size = 0;
+
+      response.on('data', (data) => {
```

Comment:
Should we reject the promise `.on('error')` (either read error or fs write error?)